WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java

```diff
@@ -1796,76 +1787,129 @@ private void trackAddMediaFromWPLibraryEvents(boolean isVideo, long mediaId) {
     }
 
-    private boolean addMediaVisualEditor(Uri uri, boolean isVideo) {
-        String path = MediaUtils.getRealPathFromURI(this, uri);
+    /*
+     * processes a list of media in the background (optimizing, resizing, etc.) and adds them to
+     * the editor one at a time
+     */
+    private class AddMediaListThread extends Thread {
+        private final List<Uri> uriList = new ArrayList<>();
+        private final Handler handler = new Handler();
+        private final boolean isNew;
+        private boolean didAnyFail;
 
-        if (path == null) {
-            ToastUtils.showToast(this, R.string.file_not_found, Duration.SHORT);
-            return false;
+        AddMediaListThread(@NonNull List<Uri> uriList, boolean isNew) {
+            this.uriList.addAll(uriList);
+            this.isNew = isNew;
+            showOverlay(true);
         }
 
-        Uri optimizedMedia = WPMediaUtils.getOptimizedMedia(this, path, isVideo);
-        if (optimizedMedia != null) {
-            uri = optimizedMedia;
-            path = MediaUtils.getRealPathFromURI(this, uri);
-        } else {
-            // Fix for the rotation issue https://github.com/wordpress-mobile/WordPress-Android/issues/5737
-            if (!mSite.isWPCom()) {
-                // If it's not wpcom we must rotate the picture locally
-                Uri rotatedMedia = WPMediaUtils.fixOrientationIssue(this, path, isVideo);
-                if (rotatedMedia != null) {
-                    uri = rotatedMedia;
-                    path = MediaUtils.getRealPathFromURI(this, uri);
+        @Override
+        public void run() {
+            for (Uri mediaUri: uriList) {
+                if (!processMedia(mediaUri)) {
+                    didAnyFail = true;
                 }
-            } else {
-                // It's a wpcom site. Just create a version of the picture rotated for the old visual editor
-                // All the other editors read EXIF data
-                if (mShowNewEditor) {
-                    Uri rotatedMedia = WPMediaUtils.fixOrientationIssue(this, path, isVideo);
+            }
+
+            handler.post(new Runnable() {
+                @Override
+                public void run() {
+                    savePostAsync(null);
+                    if (didAnyFail) {
+                        ToastUtils.showToast(EditPostActivity.this, R.string.gallery_error, ToastUtils.Duration.SHORT);
+                    }
+                    showOverlay(false);
+                }
+            });
+        }
+
+        private boolean processMedia(Uri mediaUri) {
+            if (mediaUri == null) {
+                return false;
+            }
+
+            Activity activity = EditPostActivity.this;
+
+            if (!MediaUtils.isInMediaStore(mediaUri)
+                    && !mediaUri.toString().startsWith("/")
+                    && !mediaUri.toString().startsWith("file://")) {
+                mediaUri = MediaUtils.downloadExternalMedia(activity, mediaUri);
```

Comment:
Good catch. In what situations do we need to download external media? I'm thinking that's only necessary when choosing from the device gallery, in which case we could handle it in those situations.